### PR TITLE
Perform JS static allocations at compile time

### PR DESCRIPTION
**Pull request by kripken**
First part of #7795: this removes the ability to do static allocations in JS when the app starts. That means no more `staticAlloc`, `ALLOC_STATIC`, and the `staticSealed` logic to handle it. Instead, static allocations are done at **compile** time, using `makeStaticAlloc`. The JS compiler informs `emscripten.py` of those static allocations, and then they are added after static allocations from LLVM, basically.

Benefits:

 * Remove a bunch of complexity (no more `staticSealed`), etc. See `apply_memory` in `emscripten.py` for how simply memory layout is after this: we run the C compiler, then the JS compiler, then in that python code we can lay out static memory completely at compile time.
 * Saves code size and startup work by replacing `x = staticAlloc(size)` to `x = ptr` (hardcoded ptr numbers, computed at compile time). On hello world (with closure/-Os) this saves around 1%, more than I thought actually.

Downsides:

 * The current model is very simple: JS library files do `{{{ makeStaticAlloc(size) }}}` which does the allocation and returns a ptr, at compile time. However, when JS did the call at runtime, if the code was not reached the allocation was not done. In this model, if the JS library is parsed and preprocessed, we allocate - which means the `library_pthreads.js` ones are not done by default, but the `library.js` ones are. As a result, we are wasting in some cases a little static space that is actually unused (maybe 100 bytes?). This is probably pretty negligible, but we could optimize it out later, at the cost of more complexity in the JS compiler - not sure if that would be worth it.

Possible followups / stuff not done in this PR:

 * No changes to dynamic allocation in JS when the app starts. Only static allocation is changed here.
 * We could hardcode DYNAMICTOP_PTR's location in the wasm (but not its value, not without changes to dynamic allocation, see previous point), say as a param to `wasm-emscripten-finalize` etc. This would save passing it at runtime into the wasm.

PR also contains a few minor test fixes, like the stack-overflow test had an incorrect part, but we never noticed til now where the new memory layout happens to make it actually cause a problem.

### CODE REVIEW

File: emscripten.py [hunk 1]
```diff
@@ -581,6 +590,14 @@ def read_proxied_function_signatures(asmConsts):
   if not shared.Settings.WASM_BACKEND:
     shared.Settings.PROXIED_FUNCTION_SIGNATURES = read_proxied_function_signatures(metadata['asmConsts'])
 
+  shared.Settings.STATIC_BUMP = align_static_bump(metadata)
+
+
+def apply_forwarded_data(forwarded_data):
+  forwarded_json = json.loads(forwarded_data)
+  # Be aware of JS static allocations
+  shared.Settings.STATIC_BUMP = forwarded_json['STATIC_BUMP']
```
[the remark this answers is not in the document]
Comment (kripken) [reply emscripten.py]: It is slightly odd, yeah, we use `shared.Settings` as a global location for compiler state. We do mark them in `settings.js` with comments saying some are just for internal use, like this one. But perhaps they could be in `shared.Internals` or such?

Specifically here, this can't be a local var to the file since we do want to use it elsewhere (e.g. when we write the metadata we write this - although I didn't update all those places yet in this PR, to keep it small).

File: emscripten.py [hunk 2]
```diff
@@ -596,18 +613,50 @@ def compile_settings(compiler_engine, libraries, temp_files):
...
+def apply_memory(pre, metadata):
+  # Apply the statically-at-compile-time computed memory locations.
+
+  # Memory layout:
+  #  * first the static globals
+  global_start = shared.Settings.GLOBAL_BASE
+  static_bump = shared.Settings.STATIC_BUMP
+  #  * then the stack (up on fastcomp, down on upstream)
+  stack_low = align_memory(global_start + static_bump)
+  stack_high = stack_low + shared.Settings.TOTAL_STACK
```
[the remark this answers is not in the document]
Comment (kripken) [reply emscripten.py]: Good point, I'll align that too.

I don't think we have any checks on the size of the stack, but we could have the compiler error if it's not a multiple if 16 I suppose.

File: emscripten.py [hunk 3]
```diff
@@ -621,8 +670,8 @@ def memory_and_global_initializers(pre, metadata, mem_init):
 
   if shared.Settings.SIDE_MODULE:
     pre = pre.replace('GLOBAL_BASE', 'gb')
-  if shared.Settings.SIDE_MODULE or shared.Settings.WASM:
-    pre = pre.replace('{{{ STATIC_BUMP }}}', str(staticbump))
+
+  pre = apply_memory(pre, metadata)
```
[the remark this answers is not in the document]
Comment (kripken) [reply emscripten.py]: Identical for wasm and !wasm.

In a side module we wouldn't have STACK_BASE etc., as the location is determined at runtime anyhow (by the dynamic loader). So maybe it's a little odd we compute those even though they are not written out. I'll add a comment.

File: emscripten.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply emscripten.py]: Makes sense, added that in ASSERTIONS mode at runtime, for the stack (and heap too).

**Comment by rianhunter** [comment]
This should probably increment `EMSCRIPTEN_ABI_MINOR` (while `EMSCRIPTEN_ABI_MAJOR` is still `0`).

**Comment by kripken** [comment]
@rianhunter I'm not sure this changes the ABI, yet? If we stop passing DYNAMICTOP_PTR from JS to wasm then that would be a change, but this PR doesn't go that far yet. Or is there something else here?

**Comment by rianhunter** [comment]
@kripken I looked briefly but I thought this might change the exports available to the hosted wasm. If all the same memory-layout exports are available then ignore my comment. Otherwise I understand this only changes how the JS runtime allocates static memory.

**Comment by kripken** [comment]
Ok, yeah, then this should be fine as-is. In fact I verified earlier that the wasm emitted is identical to before this PR - the change is only on the JS side.

Thanks for checking - good to have more eyes on this stuff.

File: src/library_pthread.js [hunk 4]
```diff
@@ -529,7 +528,7 @@ var LibraryPThread = {
   },
 
   _num_logical_cores__deps: ['emscripten_force_num_logical_cores'],
-  _num_logical_cores: '; if (ENVIRONMENT_IS_PTHREAD) __num_logical_cores = PthreadWorkerInit.__num_logical_cores; else { PthreadWorkerInit.__num_logical_cores = __num_logical_cores = allocate(1, "i32*", ALLOC_STATIC); HEAPU32[__num_logical_cores>>2] = navigator["hardwareConcurrency"] || ' + {{{ PTHREAD_HINT_NUM_CORES }}} + '; }',
+  _num_logical_cores: '; if (ENVIRONMENT_IS_PTHREAD) __num_logical_cores = PthreadWorkerInit.__num_logical_cores; else { PthreadWorkerInit.__num_logical_cores = __num_logical_cores = {{{ makeStaticAlloc(1) }}}; HEAPU32[__num_logical_cores>>2] = navigator["hardwareConcurrency"] || ' + {{{ PTHREAD_HINT_NUM_CORES }}} + '; }',
```
Comment (juj) [inline src/library_pthread.js]: This should be `makeStaticAlloc(4)`?

File: src/library_pthread.js [hunk 5]
```diff
@@ -1031,7 +1030,7 @@ var LibraryPThread = {
   },
 
   // Stores the memory address that the main thread is waiting on, if any.
-  _main_thread_futex_wait_address: '; if (ENVIRONMENT_IS_PTHREAD) __main_thread_futex_wait_address = PthreadWorkerInit.__main_thread_futex_wait_address; else PthreadWorkerInit.__main_thread_futex_wait_address = __main_thread_futex_wait_address = allocate(1, "i32*", ALLOC_STATIC)',
+  _main_thread_futex_wait_address: '; if (ENVIRONMENT_IS_PTHREAD) __main_thread_futex_wait_address = PthreadWorkerInit.__main_thread_futex_wait_address; else PthreadWorkerInit.__main_thread_futex_wait_address = __main_thread_futex_wait_address = {{{ makeStaticAlloc(1) }}}',
```
Comment (juj) [inline src/library_pthread.js]: `makeStaticAlloc(4)` as well?

File: src/parseTools.js [hunk 6]
```diff
@@ -1441,8 +1441,32 @@ function makeEval(code) {
...
-  size = (size + (STACK_ALIGN-1)) & -STACK_ALIGN;
-  return 'STATICTOP; STATICTOP += ' + size + ';';
+  size = alignMemory(size);
+  var ret = alignMemory(GLOBAL_BASE + STATIC_BUMP);
+  STATIC_BUMP = ret + size - GLOBAL_BASE;
+  return ret;
+}
+
+function makeStaticString(string) {
+  var len = string.length + 1;
```
Comment (juj) [inline src/parseTools.js]: This is not correct size computation for an UTF-8 string - either need to use `lengthBytesUTF8(string)` or if you want to conservatively allocate, use `string.length*4+1`.

File: tests/test_core.py [hunk 7]
```diff
@@ -3790,7 +3790,7 @@ def test_dylink_jslib(self):
   def test_dylink_global_var_jslib(self):
     create_test_file('lib.js', r'''
       mergeInto(LibraryManager.library, {
-        jslib_x: 'allocate(1, "i32*", ALLOC_STATIC)',
+        jslib_x: '{{{ makeStaticAlloc(1) }}}',
```
Comment (juj) [inline tests/test_core.py]: `makeStaticAlloc(4)`?

File: src/library_pthread.js [hunk 8]
```diff
@@ -54,7 +54,7 @@ var LibraryPThread = {
 
 #if PTHREADS_PROFILING
     createProfilerBlock: function(pthreadPtr) {
-      var profilerBlock = (pthreadPtr == PThread.mainThreadBlock) ? allocate({{{ C_STRUCTS.thread_profiler_block.__size__ }}}, "i32*", ALLOC_STATIC) : _malloc({{{ C_STRUCTS.thread_profiler_block.__size__ }}});
+      var profilerBlock = (pthreadPtr == PThread.mainThreadBlock) ? {{{ makeStaticAlloc(C_STRUCTS.thread_profiler_block.__size__) }}} : _malloc({{{ C_STRUCTS.thread_profiler_block.__size__ }}});
```
Comment (juj) [inline src/library_pthread.js]: (it looks like the above three code blocks have historically overallocated 4x the number of bytes needed, but this change fixes that)

File: src/library_pthread.js (hunk 4 above)
Comment (kripken) [reply src/library_pthread.js]: Thanks, yeah, fixing these. (Technically the allocations are stack aligned, so 1 means at least 4 anyhow, but better not to rely on that.)

File: src/parseTools.js (hunk 6 above)
Comment (kripken) [reply src/parseTools.js]: Thanks! Fixing with lengthBytesUTF8 at compile time, to not waste any space here.

File: src/settings.js [hunk 9]
```diff
@@ -1292,39 +1283,56 @@ var ASMFS = 0;
 // then you can safely ignore this warning.
 var SINGLE_FILE = 0;
 
-// For internal use only (name of the file containing wasm text, if relevant).
+// if set to 1, then generated WASM files will contain a custom
+// "emscripten_metadata" section that contains information necessary
+// to execute the file without the accompanying JS file.
+var EMIT_EMSCRIPTEN_METADATA = 0;
```
Comment (sbc100) [inline src/settings.js]: I think you added this setting my mistake here.  It was already added in #7815

Comment (kripken) [reply src/settings.js]: > I think you added this setting my mistake here. It was already added in #7815

Sorry, I don't understand what you mean here?

Comment (sbc100) [reply src/settings.js]: EMIT_EMSCRIPTEN_METADATA now appears twice in this file, after this change I think..

Comment (kripken) [reply src/settings.js]: Oh, thanks, must be a bad merge. I can clean that up unless you were already looking at stuff in that file anyhow and planned to?

Comment (sbc100) [reply src/settings.js]: I don't have a specific CL that that no
